drivers/interrupt_controller/wuc_ite_it8xxx2.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2022 ITE Corporation. All Rights Reserved
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <dt-bindings/interrupt-controller/it8xxx2-wuc.h>
+#include <kernel.h>
+#include <soc.h>
+#include <soc_common.h>
+#include <stdlib.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wuc_ite_it8xxx2, LOG_LEVEL_ERR);
+
+#define DT_DRV_COMPAT ite_it8xxx2_wuc
+
+/* WUC group from 1~16 */
+#define IT8XXX2_WUC_GROUP_COUNT		16
+
+/* WUC module instances declaration */
+static const struct device *wuc_dev[];
+
+/* Driver config */
+struct it8xxx2_wuc_cfg {
+	/* WUC wakeup edge mode register */
+	uintptr_t reg_wuemr;
+	/* WUC wakeup edge sense register */
+	uintptr_t reg_wuesr;
+	/* WUC wakeup enable register */
+	uintptr_t reg_wuenr;
+	/* WUC wakeup both edge mode register */
+	uintptr_t reg_wubemr;
+};
+
+void it8xxx2_wuc_enable(uint8_t group, uint8_t mask)
+{
+	const struct it8xxx2_wuc_cfg *config =
+		(const struct it8xxx2_wuc_cfg *)wuc_dev[(group - 1)]->config;
+	volatile uint8_t *reg_wuenr = (uint8_t *)config->reg_wuenr;
+
+	if (group > IT8XXX2_WUC_GROUP_COUNT) {
+		LOG_ERR("Invalid WUC group");
+		return;
+	}
+
+	/*
+	 * WUC group only 1, 3, and 4 have enable/disable register,
+	 * others are always enable.
+	 */
+	if (reg_wuenr == IT8XXX2_WUC_UNUSED_REG) {
+		return;
+	}
+
+	/* Enable wakeup interrupt of the pin */
+	*reg_wuenr |= mask;
+}
+
+void it8xxx2_wuc_disable(uint8_t group, uint8_t mask)
+{
+	const struct it8xxx2_wuc_cfg *config =
+		(const struct it8xxx2_wuc_cfg *)wuc_dev[(group - 1)]->config;
+	volatile uint8_t *reg_wuenr = (uint8_t *)config->reg_wuenr;
+
+	if (group > IT8XXX2_WUC_GROUP_COUNT) {
+		LOG_ERR("Invalid WUC group");
+		return;
+	}
+
+	/*
+	 * WUC group only 1, 3, and 4 have enable/disable register,
+	 * others are always enable.
+	 */
+	if (reg_wuenr == IT8XXX2_WUC_UNUSED_REG) {
+		return;
+	}
+
+	/* Disable wakeup interrupt of the pin */
+	*reg_wuenr &= ~mask;
+}
+
+void it8xxx2_wuc_clear_status(uint8_t group, uint8_t mask)
+{
+	const struct it8xxx2_wuc_cfg *config =
+		(const struct it8xxx2_wuc_cfg *)wuc_dev[(group - 1)]->config;
+	volatile uint8_t *reg_wuesr = (uint8_t *)config->reg_wuesr;
+
+	if (group > IT8XXX2_WUC_GROUP_COUNT) {
+		LOG_ERR("Invalid WUC group");
+		return;
+	}
+
+	if (reg_wuesr == IT8XXX2_WUC_UNUSED_REG) {
+		return;
+	}
+
+	/* W/C wakeup interrupt status of the pin */
+	*reg_wuesr = mask;
+}
+
+void it8xxx2_wuc_set_polarity(uint8_t group, uint8_t mask, uint32_t flags)
+{
+	const struct it8xxx2_wuc_cfg *config =
+		(const struct it8xxx2_wuc_cfg *)wuc_dev[(group - 1)]->config;
```

Comment:
If you change the first parameter to `const struct device *dev`, then you don't have to do the lookup from the `wuc_dev[]` array.

The wuc-cells only needs to be 1 to store the mask.

Reply:
I tried and got compiler warning:
'struct device' declared inside parameter list will not be visible outside of this definition or declaration in soc_common.h.

The struct device declaration is after wuc function declaration, 
so in 3rd commit I include "device.h" in soc_common.h , then the compiler shows:
error: 'ENODEV' undeclared in device.h

and

lots warning: parameter names (without types) in function declaration in all of device_extern.h functions 
ex.
    8 | extern const struct device DEVICE_DT_NAME_GET(DT_N); /* dts_ord_0 */
      |                       ^~~~~~
   9 | extern const struct device DEVICE_DT_NAME_GET(DT_N_S_aliases); /* dts_ord_1 */
      |                      ^~~~~~

Shall I look into the warnings or just use the group cell ?

Reply:
Please look into the warnings.  It's a better implementation if you don't have to specify the group cell property.

Reply:
Done